cleanup: mark google/cloud/ bazel targets as deprecated

devjgm opened the pull request:
Part of: #3701

We want users to only use targets in the top-level //BUILD.bazel file.
Very likely nobody is using these targets even right now. But a 90+ day
deprecation would be nice. We could extend this to a year if people feel
that's necessary.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8550)
<!-- Reviewable:end -->

coryan commented on google/cloud/BUILD.bazel:
```diff
@@ -12,7 +12,17 @@
+package(
+    default_deprecation = """
+    These targets are not for public use and will be removed sometime around
+    July 1, 2022. Only use Bazel targets defined in the top-level //BUILD.bazel
```
nit: 2022-07-01 is probably more readable for folks outside the US.

devjgm replied:
Good point. I'll change to that format once we know how long to deprecate these. I'd be happy to pick 2023-01-01 for all of this to keep it simple. We could also pick a full year from now. Or something else. WDYT?

coryan commented on google/cloud/grpc_utils/BUILD.bazel:
```diff
@@ -12,7 +12,17 @@
+package(
+    default_deprecation = """
+    These targets are not for public use and will be removed sometime around
+    July 1, 2022. Only use Bazel targets defined in the top-level //BUILD.bazel
```
The only target here is slated for deprecation in 2023-01-01 (see around line 41), it seems like we should stick to that date.

coryan commented on google/cloud/BUILD.bazel:
```diff
@@ -12,7 +12,17 @@
-package(default_visibility = ["//visibility:public"])
+# TODO(#3701) Change the default_visibility to "//visibility:private", and
+# widen visibility as needed with per-target visibility.
+package(
```
This file happens to have some libraries that have (AFAIK) no corresponding top-level target: `google_cloud_cpp_common` and `google_cloud_cpp_grpc_utils`.

Maybe for them we should decide if we are going to make these libraries public (I think we should).  And I think the deprecation period may need to be longer?

coryan reviewed (changes requested):
There are some `*_testing` libraries that get deprecated with this change.  I think they should be deprecated, maybe the time period is too short.

This also deprecates the only want to link directly against `google_cloud_cpp_common` and `google_cloud_cpp_grpc_utils`.  Those seem to be part of our public API. We should add top-level targets for them.  The deprecation period also feels short for them.

coryan commented on google/cloud/testing_util/BUILD.bazel:
```diff
@@ -12,7 +12,17 @@
+    more details.
+    """,
+    default_visibility = ["//visibility:public"],
+)
```
Argh, also here, there are rules here that are already deprecated.

devjgm replied:
Yes, I saw those. But to keep things simpler, I was leaning toward just handling these deprecations as whole files, even if an already deprecated thing stays around a bit longer. I could remove the other deprecation notices though.

coryan replied:
Delaying the deprecation to simplify things SGTM.  Removing them right away also SGTM.

devjgm commented:
> This also deprecates the only want to link directly against `google_cloud_cpp_common` and `google_cloud_cpp_grpc_utils`. Those seem to be part of our public API. We should add top-level targets for them. The deprecation period also feels short for them.

I was debating whether "common" was actually part of our public API, or if the intent is that you just get them as part of your dep on one of our public user-facing libraries. If they are intended to be part of our public API on their own, then I agree we need top-level targets for them. However, it's not 100% clear to me that we want that.

coryan commented:
> I was debating whether "common" was actually part of our public API, or if the intent is that you just get them as part of your dep on one of our public user-facing libraries. If they are intended to be part of our public API on their own, then I agree we need top-level targets for them. However, it's not 100% clear to me that we want that.

I think the ship named "Should we make this part of our public API" sailed, picked up cargo in 5 ports, came back, delivered the cargo, and sailed again. 😄

What follows is not as black and white, but arguably, there are APIs in these libraries useful when using C++ in GCP, and can be used even without any client library, e.g.:
  - `g::c::Project`, `g::c::KmsKeyName` (and I expect we will add `g::c::Location` at some point).
  - `g::c::future` and `g::c::CompletionQueue` might be useful outside the libraries too.

devjgm commented:
> > I was debating whether "common" was actually part of our public API, or if the intent is that you just get them as part of your dep on one of our public user-facing libraries. If they are intended to be part of our public API on their own, then I agree we need top-level targets for them. However, it's not 100% clear to me that we want that.
> 
> I think the ship named "Should we make this part of our public API" sailed, picked up cargo in 5 ports, came back, delivered the cargo, and sailed again. 😄

For the C++ code, I agree. For the build targets, I don't think it's clear, but I'm happy to defer to your preference for how to proceed. I'll prepare a PR to add them to //BUILD.bazel.

> 
> What follows is not as black and white, but arguably, there are APIs in these libraries useful when using C++ in GCP, and can be used even without any client library, e.g.:
> 
> * `g::c::Project`, `g::c::KmsKeyName` (and I expect we will add `g::c::Location` at some point).
> * `g::c::future` and `g::c::CompletionQueue` might be useful outside the libraries too.

devjgm commented:
@coryan PTAL.

Some of the noteworthy changes.
* The bazel config_settings that were already deprecated I removed in https://github.com/googleapis/google-cloud-cpp/pull/8556
* I pushed back the deprecation date a full year to 2023-03-16, which should cover us in all these cases.
* I changed the date format to something more universal.
* I exposed the common and grpc_utils targets in the top-level BUILD file in https://github.com/googleapis/google-cloud-cpp/pull/8552
